File: packages/rust-sdk/src/contacts.rs
use std::sync::Arc;

use reqwest::Method;

use crate::{Config, Result, list_opts::{ListOptions, ListResponse}};
use crate::contacts::types::{Contact, ContactResponse, CreateContactRequest, SearchContactRequest, UpdateContactRequest};

/// Kibamail APIs for `/contacts` endpoints.
#[derive(Clone, Debug)]
pub struct ContactsSvc(pub(crate) Arc<Config>);

impl ContactsSvc {
    /// Create a new contact in your workspace.
    ///
    /// <https://kibamail.com/docs/api-reference/contacts/create-contact>
    #[maybe_async::maybe_async]
    #[allow(clippy::needless_pass_by_value)]
    pub async fn create(
        &self,
        params: impl Into<CreateContactRequest>,
    ) -> Result<ContactResponse> {
        let params = params.into();
        let request = self.0.build(Method::POST, "v1/contacts");
        let response = self.0.send(request.json(&params)).await?;
        let content = response.json::<ContactResponse>().await?;

        Ok(content)
    }

    /// Retrieve a paginated list of all contacts in your workspace.
    ///
    /// <https://kibamail.com/docs/api-reference/contacts/list-contacts>
    #[maybe_async::maybe_async]
    #[allow(clippy::needless_pass_by_value)]
    pub async fn list(&self, params: impl Into<ListOptions>) -> Result<ListResponse<Contact>> {
        let params = params.into();
        let request = self.0.build(Method::GET, "v1/contacts").query(&params);
        let response = self.0.send(request).await?;
        let content = response.json::<ListResponse<Contact>>().await?;

        Ok(content)
    }

    /// Retrieve a specific contact by ID.
    ///
    /// <https://kibamail.com/docs/api-reference/contacts/get-contact>
    #[maybe_async::maybe_async]
    pub async fn get(&self, contact_id: &str) -> Result<Contact> {
        let path = format!("v1/contacts/{contact_id}");
        let request = self.0.build(Method::GET, &path);
        let response = self.0.send(request).await?;
        let content = response.json::<Contact>().await?;

        Ok(content)
    }

    /// Update an existing contact by ID.
    ///
    /// <https://kibamail.com/docs/api-reference/contacts/update-contact>
    #[maybe_async::maybe_async]
    #[allow(clippy::needless_pass_by_value)]
    pub async fn update(
        &self,
        contact_id: &str,
        params: impl Into<UpdateContactRequest>,
    ) -> Result<ContactResponse> {
        let params = params.into();
        let path = format!("v1/contacts/{contact_id}");
        let request = self.0.build(Method::PUT, &path);
        let response = self.0.send(request.json(&params)).await?;
        let content = response.json::<ContactResponse>().await?;

        Ok(content)
    }

    /// Delete a contact by ID.
    ///
    /// <https://kibamail.com/docs/api-reference/contacts/delete-contact>
    #[maybe_async::maybe_async]
    pub async fn delete(&self, contact_id: &str) -> Result<()> {
        let path = format!("v1/contacts/{contact_id}");
        let request = self.0.build(Method::DELETE, &path);
        self.0.send(request).await?;

        Ok(())
    }

    /// Search contacts using advanced filtering conditions.
    ///
    /// <https://kibamail.com/docs/api-reference/contacts/search-contacts>
    #[maybe_async::maybe_async]
    #[allow(clippy::needless_pass_by_value)]
    pub async fn search(
        &self,
        params: impl Into<SearchContactRequest>,
    ) -> Result<ListResponse<Contact>> {
        let params = params.into();
        let request = self.0.build(Method::POST, "v1/contacts/search");
        let response = self.0.send(request.json(&params)).await?;
        let content = response.json::<ListResponse<Contact>>().await?;

        Ok(content)
    }
}

#[allow(unreachable_pub)]
pub mod types {
    use std::collections::HashMap;

    use serde::{Deserialize, Serialize};

    crate::define_id_type!(ContactId);

    /// Request to create a new contact.
    #[must_use]
    #[derive(Debug, Clone, Default, Serialize, Deserialize)]
    #[serde(rename_all = "camelCase")]
    pub struct CreateContactRequest {
        /// Contact email address (required).
        pub email: String,
        /// Contact first name.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub first_name: Option<String>,
        /// Contact last name.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub last_name: Option<String>,
        /// Contact phone number.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub phone: Option<String>,
        /// Contact country.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub country: Option<String>,
        /// Contact city.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub city: Option<String>,
        /// Contact timezone.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub timezone: Option<String>,
        /// Custom properties for the contact.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub properties: Option<HashMap<String, serde_json::Value>>,
        /// Topic IDs to subscribe the contact to.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub topics: Option<Vec<String>>,
    }

    /// Request to update an existing contact.
    #[must_use]
    #[derive(Debug, Clone, Default, Serialize, Deserialize)]
    #[serde(rename_all = "camelCase")]
    pub struct UpdateContactRequest {
        /// Contact email address.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub email: Option<String>,
        /// Contact first name.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub first_name: Option<String>,
        /// Contact last name.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub last_name: Option<String>,
        /// Contact phone number.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub phone: Option<String>,
        /// Contact country.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub country: Option<String>,
        /// Contact city.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub city: Option<String>,
        /// Contact timezone.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub timezone: Option<String>,
        /// Custom properties for the contact.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub properties: Option<HashMap<String, serde_json::Value>>,
        /// Topic IDs to subscribe the contact to.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub topics: Option<Vec<String>>,
    }

    /// Request to search contacts with filtering conditions.
    #[must_use]
    #[derive(Debug, Clone, Serialize, Deserialize)]
    #[serde(rename_all = "camelCase")]
    pub struct SearchContactRequest {
        /// Search conditions (can be complex nested conditions).
        pub conditions: serde_json::Value,
        /// Maximum number of items to return.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub limit: Option<u32>,
        /// Cursor for fetching items after this cursor.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub after: Option<String>,
        /// Cursor for fetching items before this cursor.
        #[serde(skip_serializing_if = "Option::is_none")]
        pub before: Option<String>,
    }

    /// Contact object.
    #[must_use]
    #[derive(Debug, Clone, Deserialize, Serialize)]
    #[serde(rename_all = "camelCase")]
    pub struct Contact {
        /// Unique contact identifier.
        pub id: ContactId,
        /// Contact email address.
        pub email: String,
        /// Contact first name.
        #[serde(default)]
        pub first_name: String,
        /// Contact last name.
        #[serde(default)]
        pub last_name: String,
        /// Contact phone number.
        #[serde(default)]
        pub phone: String,
        /// Contact country.
        #[serde(default)]
        pub country: String,
        /// Contact city.
        #[serde(default)]
        pub city: String,
        /// Contact timezone.
        #[serde(default)]
        pub timezone: String,
        /// Contact subscription status.
        pub status: String,
        /// Custom properties.
        #[serde(default)]
        pub properties: HashMap<String, serde_json::Value>,
        /// Subscribed topic IDs.
        #[serde(default)]
        pub topics: Vec<String>,
        /// Timestamp when contact was created.
        #[serde(default)]
        pub created_at: String,
        /// Timestamp when contact was last updated.
        #[serde(default)]
        pub updated_at: String,
    }

    /// Response from creating or updating a contact.
    #[must_use]
    #[derive(Debug, Clone, Deserialize, Serialize)]
    pub struct ContactResponse {
        /// The ID of the created/updated contact.
        pub id: ContactId,
    }
}

#[cfg(test)]
mod test {
    use crate::test::CLIENT;
    use crate::types::{CreateContactRequest, UpdateContactRequest, SearchContactRequest};
    use crate::{Result, list_opts::ListOptions};
    use std::collections::HashMap;

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_create_contact() -> Result<()> {
        let client = &*CLIENT;

        let result = client
            .contacts
            .create(CreateContactRequest {
                email: "john.doe@example.com".to_string(),
                first_name: Some("John".to_string()),
                last_name: Some("Doe".to_string()),
                ..Default::default()
            })
            .await?;

        assert!(!result.id.is_empty());

        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_create_contact_with_properties() -> Result<()> {
        let client = &*CLIENT;

        let result = client
            .contacts
            .create(CreateContactRequest {
                email: "jane.smith@example.com".to_string(),
                first_name: Some("Jane".to_string()),
                last_name: Some("Smith".to_string()),
                properties: Some(HashMap::from([
                    ("Company".to_string(), serde_json::json!("Acme Inc")),
                    ("Plan".to_string(), serde_json::json!("Enterprise")),
                ])),
                ..Default::default()
            })
            .await?;

        assert!(!result.id.is_empty());

        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_create_contact_with_topics() -> Result<()> {
        let client = &*CLIENT;

        let result = client
            .contacts
            .create(CreateContactRequest {
                email: "subscriber@example.com".to_string(),
                first_name: Some("New".to_string()),
                last_name: Some("Subscriber".to_string()),
                topics: Some(vec!["topic_newsletter".to_string(), "topic_updates".to_string()]),
                ..Default::default()
            })
            .await?;

        assert!(!result.id.is_empty());

        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_list_contacts() -> Result<()> {
        let client = &*CLIENT;

        let _result = client.contacts.list(ListOptions::default()).await?;

        // Successfully listed contacts (may be empty)
        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_list_contacts_with_pagination() -> Result<()> {
        let client = &*CLIENT;

        let _result = client
            .contacts
            .list(ListOptions::new().with_limit(10))
            .await?;

        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_get_contact() -> Result<()> {
        let client = &*CLIENT;

        // Create a contact first
        let created = client
            .contacts
            .create(CreateContactRequest {
                email: "get.test@example.com".to_string(),
                first_name: Some("Get".to_string()),
                last_name: Some("Test".to_string()),
                ..Default::default()
            })
            .await?;

        // Get the contact
        let _contact = client.contacts.get(&created.id).await?;

        // Successfully retrieved contact
        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_update_contact() -> Result<()> {
        let client = &*CLIENT;

        // Create a contact first
        let created = client
            .contacts
            .create(CreateContactRequest {
                email: "update.test@example.com".to_string(),
                first_name: Some("Update".to_string()),
                last_name: Some("Test".to_string()),
                ..Default::default()
            })
            .await?;

        // Update the contact
        let updated = client
            .contacts
            .update(
                &created.id,
                UpdateContactRequest {
                    first_name: Some("Updated".to_string()),
                    last_name: Some("Name".to_string()),
                    ..Default::default()
                },
            )
            .await?;

        assert_eq!(updated.id, created.id);

        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_delete_contact() -> Result<()> {
        let client = &*CLIENT;

        // Create a contact first
        let created = client
            .contacts
            .create(CreateContactRequest {
                email: "delete.test@example.com".to_string(),
                first_name: Some("Delete".to_string()),
                last_name: Some("Test".to_string()),
                ..Default::default()
            })
            .await?;

        // Delete the contact
        client.contacts.delete(&created.id).await?;

        Ok(())
    }

    #[tokio::test]
    #[cfg(not(feature = "blocking"))]
    async fn test_search_contacts() -> Result<()> {
        let client = &*CLIENT;

        let _result = client
            .contacts
            .search(SearchContactRequest {
                conditions: serde_json::json!({
                    "AND": [
                        { "field": "status", "operator": "equals", "value": "SUBSCRIBED" }
                    ]
                }),
                limit: Some(50),
                after: None,
                before: None,
            })
            .await?;

        Ok(())
    }
}

File: sdks/nodejs/resources/contacts.ts
import type { createHttpClient } from "../client";
import type { paths } from "../schema";

type HttpClient = ReturnType<typeof createHttpClient>;

// Extract types from schema
type CreateContactBody = paths["/v1/contacts"]["post"]["requestBody"]["content"]["application/json"];
type UpdateContactBody = paths["/v1/contacts/{contactId}"]["put"]["requestBody"]["content"]["application/json"];
type ListContactsQuery = paths["/v1/contacts"]["get"]["parameters"]["query"];
type SearchContactsBody = paths["/v1/contacts/search"]["post"]["requestBody"]["content"]["application/json"];

/**
 * Contacts Resource
 *
 * Manage contacts in your workspace. Contacts are the core entities that receive your emails
 * and represent your subscribers, users, or customers.
 *
 * **Key Features:**
 * - Create and manage contact profiles with built-in and custom properties
 * - Subscribe contacts to topics for targeted communication
 * - Search contacts using advanced filtering conditions
 * - Track subscription status and preferences
 * - Cursor-based pagination for efficient data retrieval
 *
 * @example
 * ```ts
 * const kibamail = new Kibamail("your-api-key");
 *
 * // Create a contact
 * const contact = await kibamail.contacts.create({
 *   email: "user@example.com",
 *   firstName: "John",
 *   lastName: "Doe",
 *   properties: {
 *     "Company": "Acme Inc",
 *     "Plan": "Enterprise"
 *   }
 * });
 *
 * // List all contacts
 * const contacts = await kibamail.contacts.list({ limit: 50 });
 *
 * // Search for contacts
 * const results = await kibamail.contacts.search({
 *   filters: {
 *     AND: [
 *       { field: "Company", operator: "equals", value: "Acme Inc" },
 *       { field: "status", operator: "equals", value: "SUBSCRIBED" }
 *     ]
 *   }
 * });
 * ```
 */
export class Contacts {
  constructor(protected client: HttpClient) {}

  /**
   * Create a new contact in your workspace.
   *
   * Creates a contact record with optional custom properties and topic subscriptions.
   * If a contact with the same email already exists, this will fail with a conflict error.
   *
   * **Use Cases:**
   * - Add new subscribers from your application signup flow
   * - Import contacts from external systems or CSV files
   * - Create contact records from form submissions
   * - Sync users from your database to Kibamail
   *
   * **Behavior:**
   * - Email must be unique within the workspace
   * - Custom properties are mapped by name (must be pre-defined in workspace)
   * - Topics array automatically subscribes contact to specified topics
   * - Returns the contact ID immediately after creation
   * - Contact status defaults to 'SUBSCRIBED' unless specified
   *
   * **Required Scope:** `write:contacts`
   *
   * @param params - Contact creation parameters
   * @param params.email - Contact's email address (required, must be unique)
   * @param params.firstName - Contact's first name
   * @param params.lastName - Contact's last name
   * @param params.phone - Contact's phone number
   * @param params.country - Contact's country code (ISO 3166-1 alpha-2, e.g., "US", "GB")
   * @param params.city - Contact's city
   * @param params.timezone - Contact's timezone (e.g., "America/New_York", "Europe/London")
   * @param params.properties - Custom properties as key-value pairs (property names must exist in workspace)
   * @param params.topics - Array of topic IDs to subscribe the contact to
   *
   * @returns Promise containing the created contact ID
   *
   * @throws {BadRequestError} Invalid input data or unknown property names
   * @throws {ConflictError} Contact with this email already exists
   * @throws {UnauthorizedError} Invalid API key or missing write:contacts scope
   *
   * @example
   * ```ts
   * // Basic contact creation
   * const contact = await kibamail.contacts.create({
   *   email: "john@example.com",
   *   firstName: "John",
   *   lastName: "Doe"
   * });
   *
   * // Contact with custom properties and topics
   * const advancedContact = await kibamail.contacts.create({
   *   email: "sarah@example.com",
   *   firstName: "Sarah",
   *   lastName: "Smith",
   *   country: "US",
   *   timezone: "America/Los_Angeles",
   *   properties: {
   *     "Company": "Acme Corp",
   *     "Plan": "Enterprise",
   *     "Sign Up Date": Date.now(),
   *     "Monthly Spend": 499.99
   *   },
   *   topics: ["topic_123", "topic_456"] // Subscribe to marketing and product updates
   * });
   * ```
   */
  create(params: CreateContactBody) {
    return this.client.POST("/v1/contacts", {
      body: params,
    });
  }

  /**
   * Retrieve a paginated list of all contacts in your workspace.
   *
   * Returns contacts in reverse chronological order (newest first) with cursor-based pagination
   * for efficient traversal of large contact lists.
   *
   * **Use Cases:**
   * - Export your entire contact database
   * - Build custom contact management dashboards
   * - Sync contacts with external CRM systems
   * - Generate contact reports and analytics
   * - Display contact lists in your application
   *
   * **Behavior:**
   * - Returns contacts in reverse chronological order (newest first)
   * - Uses cursor-based pagination for efficient data retrieval
   * - Includes all contact properties (default and custom)
   * - Includes subscription status and topic memberships
   * - Maximum 100 contacts per request (use 'limit' parameter)
   * - Deleted contacts are not included in results
   *
   * **Required Scope:** `read:contacts`
   *
   * **Pagination:**
   * - Use 'after' cursor to fetch the next page
   * - Use 'before' cursor to fetch the previous page
   * - Response includes 'hasMore' and 'hasPrevious' flags
   * - Cursors are based on contact IDs for stable pagination
   *
   * **Performance Tips:**
   * - Use the maximum limit (100) for bulk exports
   * - Cache results when possible to reduce API calls
   * - For filtered results, use the search() method instead
   *
   * @param params - Pagination parameters
   * @param params.limit - Number of contacts to return (default: 20, max: 100)
   * @param params.after - Cursor for pagination - ID of the last contact from the previous page
   * @param params.before - Cursor for pagination - ID of the first contact from the next page
   *
   * @returns Promise containing paginated contact list
   *
   * @throws {UnauthorizedError} Invalid API key or missing read:contacts scope
   *
   * @example
   * ```ts
   * // Get first page of contacts
   * const firstPage = await kibamail.contacts.list({ limit: 50 });
   * console.log(firstPage.data); // Array of contacts
   * console.log(firstPage.hasMore); // true if more contacts exist
   *
   * // Get next page using cursor
   * if (firstPage.hasMore) {
   *   const lastContact = firstPage.data[firstPage.data.length - 1];
   *   const nextPage = await kibamail.contacts.list({
   *     limit: 50,
   *     after: lastContact.id
   *   });
   * }
   *
   * // Iterate through all contacts
   * let cursor = undefined;
   * let allContacts = [];
   * do {
   *   const page = await kibamail.contacts.list({ limit: 100, after: cursor });
   *   allContacts.push(...page.data);
   *   cursor = page.hasMore ? page.data[page.data.length - 1].id : undefined;
   * } while (cursor);
   * ```
   */
  list(params?: ListContactsQuery) {
    return this.client.GET("/v1/contacts", {
      params: {
        query: params,
      },
    });
  }

  /**
   * Retrieve a specific contact by ID.
   *
   * Returns the full contact record including all built-in fields and custom properties.
   *
   * **Use Cases:**
   * - Fetch contact details for display in your application
   * - Verify contact information before sending emails
   * - Retrieve contact data for analytics or reporting
   * - Get the latest contact status and properties
   *
   * **Behavior:**
   * - Returns contact only if it belongs to the authenticated workspace
   * - Includes all custom properties mapped by name
   * - Returns 404 if contact doesn't exist or belongs to different workspace
   *
   * **Required Scope:** `read:contacts`
   *
   * @param contactId - The unique identifier of the contact
   *
   * @returns Promise containing the contact details
   *
   * @throws {NotFoundError} Contact not found or doesn't belong to your workspace
   * @throws {UnauthorizedError} Invalid API key or missing read:contacts scope
   *
   * @example
   * ```ts
   * const contact = await kibamail.contacts.get("contact_abc123");
   * console.log(contact.data.email); // "user@example.com"
   * console.log(contact.data.properties); // { "Company": "Acme Inc", ... }
   * console.log(contact.data.status); // "SUBSCRIBED" | "UNSUBSCRIBED"
   * ```
   */
  get(contactId: string) {
    return this.client.GET("/v1/contacts/{contactId}", {
      params: {
        path: { contactId },
      },
    });
  }

  /**
   * Update an existing contact by ID.
   *
   * Updates contact information including built-in fields, custom properties, and topic subscriptions.
   * Only provided fields are updated; omitted fields remain unchanged.
   *
   * **Use Cases:**
   * - Update contact information when users change their profile
   * - Add or modify custom property values
   * - Change contact subscription status
   * - Update topic subscriptions
   * - Sync contact data from your application database
   *
   * **Behavior:**
   * - Only updates fields that are provided in the request
   * - Omitted fields remain unchanged
   * - Properties object merges with existing properties (provide property to update it)
   * - Topics array completely replaces existing subscriptions when provided
   * - Returns the contact ID after successful update
   *
   * **Required Scope:** `write:contacts`
   *
   * @param contactId - The unique identifier of the contact to update
   * @param params - Contact update parameters (all optional, only provided fields are updated)
   * @param params.email - New email address (must be unique)
   * @param params.firstName - Updated first name
   * @param params.lastName - Updated last name
   * @param params.phone - Updated phone number
   * @param params.country - Updated country code
   * @param params.city - Updated city
   * @param params.timezone - Updated timezone
   * @param params.properties - Custom properties to update (property names must exist in workspace)
   * @param params.topics - New topic subscriptions (completely replaces existing subscriptions)
   *
   * @returns Promise containing the updated contact ID
   *
   * @throws {NotFoundError} Contact not found or doesn't belong to your workspace
   * @throws {BadRequestError} Invalid input data or unknown property names
   * @throws {ConflictError} Email already in use by another contact
   * @throws {UnauthorizedError} Invalid API key or missing write:contacts scope
   *
   * @example
   * ```ts
   * // Update basic contact info
   * await kibamail.contacts.update("contact_abc123", {
   *   firstName: "Jane",
   *   lastName: "Smith"
   * });
   *
   * // Update custom properties
   * await kibamail.contacts.update("contact_abc123", {
   *   properties: {
   *     "Plan": "Pro",
   *     "Monthly Spend": 99.99
   *   }
   * });
   *
   * // Update topic subscriptions (replaces all existing subscriptions)
   * await kibamail.contacts.update("contact_abc123", {
   *   topics: ["topic_newsletter", "topic_product_updates"]
   * });
   *
   * // Remove all topic subscriptions
   * await kibamail.contacts.update("contact_abc123", {
   *   topics: []
   * });
   * ```
   */
  update(contactId: string, params: UpdateContactBody) {
    return this.client.PUT("/v1/contacts/{contactId}", {
      params: {
        path: { contactId },
      },
      body: params,
    });
  }

  /**
   * Delete a contact by ID.
   *
   * Permanently removes a contact from your workspace. This action cannot be undone.
   *
   * **Use Cases:**
   * - Process GDPR/privacy deletion requests
   * - Remove invalid or test contacts
   * - Clean up duplicate contacts
   * - Remove contacts who requested account deletion
   *
   * **Behavior:**
   * - Permanently deletes the contact record
   * - Also removes all associated topic subscriptions
   * - Cannot be undone - contact data is permanently lost
   * - Returns 404 if contact doesn't exist
   *
   * **Required Scope:** `write:contacts`
   *
   * **Warning:** This operation is irreversible. Consider unsubscribing contacts instead of
   * deleting them if you may need to restore them later.
   *
   * @param contactId - The unique identifier of the contact to delete
   *
   * @returns Promise containing the deleted contact ID
   *
   * @throws {NotFoundError} Contact not found or doesn't belong to your workspace
   * @throws {UnauthorizedError} Invalid API key or missing write:contacts scope
   *
   * @example
   * ```ts
   * // Delete a contact
   * await kibamail.contacts.delete("contact_abc123");
   *
   * // Delete with error handling
   * try {
   *   await kibamail.contacts.delete("contact_abc123");
   *   console.log("Contact deleted successfully");
   * } catch (error) {
   *   if (error.code === "CONTACT_NOT_FOUND") {
   *     console.log("Contact was already deleted");
   *   }
   * }
   * ```
   */
  delete(contactId: string) {
    return this.client.DELETE("/v1/contacts/{contactId}", {
      params: {
        path: { contactId },
      },
    });
  }

  /**
   * Search contacts using advanced filtering conditions.
   *
   * Search for contacts that match specific criteria using MongoDB-style query conditions.
   * Supports complex queries with AND/OR logic and various comparison operators.
   *
   * **Use Cases:**
   * - Find contacts that match specific criteria (e.g., all Enterprise customers)
   * - Build dynamic segments based on contact properties
   * - Filter contacts for targeted email campaigns
   * - Generate reports based on contact attributes
   * - Find contacts for automation workflows
   *
   * **Behavior:**
   * - Uses MongoDB-style query syntax for filters
   * - Supports AND/OR logical operators for complex queries
   * - Can filter on built-in fields and custom properties
   * - Returns paginated results with cursor-based pagination
   * - Maximum 100 contacts per request
   *
   * **Required Scope:** `read:contacts`
   *
   * **Supported Operators:**
   * - `equals`: Exact match (e.g., status equals "SUBSCRIBED")
   * - `not_equals`: Does not match
   * - `contains`: String contains value (case-insensitive)
   * - `not_contains`: String does not contain value
   * - `greater_than`: Numeric/date comparison (e.g., signup date > timestamp)
   * - `less_than`: Numeric/date comparison
   * - `greater_than_or_equal`: Numeric/date comparison
   * - `less_than_or_equal`: Numeric/date comparison
   * - `in`: Value is in array (e.g., status in ["SUBSCRIBED", "PENDING"])
   * - `not_in`: Value is not in array
   *
   * @param params - Search parameters
   * @param params.filters - Query conditions using AND/OR logic
   * @param params.limit - Number of contacts to return (default: 20, max: 100)
   * @param params.after - Cursor for pagination
   * @param params.before - Cursor for pagination
   *
   * @returns Promise containing paginated search results
   *
   * @throws {BadRequestError} Invalid filter conditions or unknown property names
   * @throws {UnauthorizedError} Invalid API key or missing read:contacts scope
   *
   * @example
   * ```ts
   * // Find all subscribed contacts in the US
   * const usContacts = await kibamail.contacts.search({
   *   filters: {
   *     AND: [
   *       { field: "status", operator: "equals", value: "SUBSCRIBED" },
   *       { field: "country", operator: "equals", value: "US" }
   *     ]
   *   }
   * });
   *
   * // Find Enterprise customers who signed up in the last 30 days
   * const recentEnterpriseCustomers = await kibamail.contacts.search({
   *   filters: {
   *     AND: [
   *       { field: "Plan", operator: "equals", value: "Enterprise" },
   *       { field: "Sign Up Date", operator: "greater_than", value: Date.now() - 30 * 24 * 60 * 60 * 1000 }
   *     ]
   *   },
   *   limit: 100
   * });
   *
   * // Find contacts with complex conditions (Enterprise OR Pro plan, AND subscribed)
   * const premiumContacts = await kibamail.contacts.search({
   *   filters: {
   *     AND: [
   *       {
   *         OR: [
   *           { field: "Plan", operator: "equals", value: "Enterprise" },
   *           { field: "Plan", operator: "equals", value: "Pro" }
   *         ]
   *       },
   *       { field: "status", operator: "equals", value: "SUBSCRIBED" }
   *     ]
   *   }
   * });
   *
   * // Find contacts with email containing specific domain
   * const companyContacts = await kibamail.contacts.search({
   *   filters: {
   *     AND: [
   *       { field: "email", operator: "contains", value: "@acme.com" }
   *     ]
   *   }
   * });
   * ```
   */
  search(params: SearchContactsBody) {
    return this.client.POST("/v1/contacts/search", {
      body: params,
    });
  }
}
